sharedbuffers/mapped_struct.py

```diff
@@ -4265,251 +4590,159 @@ def sorted_contains(a, hkey):
     return ix < hi
 
 if cython.compiled:
-    #@cython.cfunc
+
+    @cython.nogil
+    @cython.cfunc
+    @cython.returns(cython.size_t)
     @cython.locals(
         length1 = cython.size_t, length2 = cython.size_t, destlength = cython.size_t,
-        stride0 = cython.size_t, ref = cython.ulonglong,
+        stride0 = cython.size_t,
         pindex1 = cython.p_char, pindex2 = cython.p_char, pdest = cython.p_char,
-        pend1 = cython.p_char, pend2 = cython.p_char, pdestend = cython.p_char, pdeststart = cython.p_char)
-    #@cython.returns(cython.size_t)
-    def _c_merge_ui64(pindex1, length1, pindex2, length2, pdest, destlength, stride0):
+        pend1 = cython.p_char, pend2 = cython.p_char, pdestend = cython.p_char, pdeststart = cython.p_char,
+        ref = numeric_A)
+    def _c_merge_gen(pindex1, length1, pindex2, length2, pdest, destlength, stride0, ref):
         # Main merge
         pend1 = pindex1 + stride0 * length1
         pend2 = pindex2 + stride0 * length2
         pdestend = pdest + stride0 * destlength
         pdeststart = pdest
         while pindex1 < pend1 and pindex2 < pend2 and pdest < pdestend:
-            ref = cython.cast(cython.p_ulonglong, pindex2)[0]
-            while pindex1 < pend1 and cython.cast(cython.p_ulonglong, pindex1)[0] <= ref and pdest < pdestend:
-                cython.cast(cython.p_ulonglong, pdest)[0] = cython.cast(cython.p_ulonglong, pindex1)[0]
-                cython.cast(cython.p_ulonglong, pdest)[1] = cython.cast(cython.p_ulonglong, pindex1)[1]
+            ref = cython.cast('numeric_A *', pindex2)[0]
```

Comment:
Well, remember that `ref` is a fused-type parameter, and thus, it's only used to have cython deduce the correct type. It's not useful in that sense. But, since I need a variable of that very type, I might as well use it ;)